File: example/GBAnisotropy/madePolyAnisotropyInformation.py
import numpy as np

def generate_matrix(dim, v1, v2, v3):
    v = np.zeros((3 * dim, dim))
    for i in range(3 * dim):
        if i < dim:
            for j in range(dim):
                v[i, j] = v1
        elif i < 2 * dim:
            for j in range(dim):
                v[i, j] = v2
        else:
            for j in range(dim):
                v[i, j] = v3
    return v

def change_value(matrix, index, i, j, value):
    dim = matrix.shape[1]
    matrix[i + dim * (index - 1) - 1, j - 1] = value
    matrix[j + dim * (index - 1) - 1, i - 1] = value

def write(filename, variable):
    file = open(filename, "w")
    file.write(">>>>>>>>> GB properties <<<<<<<<<\n")
    file.write("---GB energy (sigma_ij), mobility prefactor (mob0_ij), activation energy (Q_ij)---\n")
    shape = variable.shape
    for i in range(shape[0]):
        for j in range(shape[1]):
            if i >= shape[1] and i < 2 * shape[1]:
                file.write(f"{variable[i, j]:.2e} ")
            else:
                file.write(f"{variable[i, j]:.3f} ")
        file.write(f"\n")
    file.close()

if __name__ == "__main__":
    filename = "./poly_anisotropy_mobility.txt"
    v = generate_matrix(15, 0.708, 2.5e-6, 0.23)
    change_value(v, 2, 1, 3, 5.0e-6) # change_value(matrix, index, i, j, value)
    change_value(v, 2, 4, 5, 5.0e-6) # change_value(matrix, index, i, j, value)
    write(filename, v)

File: example/GBAnisotropy/made_poly_anisotropy_information.m
% 用于生成GBAnisotropy中的晶界能、晶界迁移率和晶界激活能
% 生成txt文件
%------------------参数处理-------------------------------

count = input('please input the number of grain:')
V=zeros(3*count,count);

if i = 1:3*count
    if 0<i<=count
        for j = 1:count
            V(i,j) = 0.708;
        end
    elseif count< i <= 2*count
        for j = 1:count
            V(i,j) = 0.708;
        end
    else
        for j = 1:count
            V(i,j) = 0.23;
        end


%-----------------------------输出---------------------------------------
% filename = strcat('grn_',num2str(count),'_rand_2D.tex')
filename = 'poly_anisotropy_mobility.txt'
f=fullfile('D:\Matlab2021A\bin\inputfile',filename);
[r,c]=size(V);
fid=fopen(filename,'w');
fprintf(fid,'>>>>>>>>> GB properties <<<<<<<<<\n');
% fprintf(fid,'\n');
fprintf(fid,'---GB energy (sigma_ij), mobility prefactor (mob0_ij), activation energy (Q_ij)---\n');
% fprintf(fid,'B ');
% fprintf(fid,'%d\n',r);

for i=1:r
    fprintf(fid,'   ');
    for j=1:c
        if j==c
            fprintf(fid,'%3.2f\n',V(i,j));%如果是最后一个，就换行
        else
            if V(i,j)>0 && V(i,j)<10
               fprintf(fid,' '); 
            end
            fprintf(fid,'%4.2f',V(i,j));%如果不是最后一个，就tab
            fprintf(fid,'   ');
        end
    end

end
fclose(fid);
